lib/services/local/users.go

```diff
@@ -305,66 +367,115 @@ func (s *IdentityService) DeleteWebSession(user, sid string) error {
 	return err
 }
 
-// UpsertPassword upserts new password and HOTP token
-func (s *IdentityService) UpsertPassword(user string,
-	password []byte) (hotpURL string, hotpQR []byte, err error) {
-
+// UpsertPassword upserts new password and OTP token and returns OTP url and QR code.
+func (s *IdentityService) UpsertPassword(user string, password []byte) (otpURL string, otpQRCode []byte, err error) {
 	if err := services.VerifyPassword(password); err != nil {
 		return "", nil, err
 	}
+
 	hash, err := bcrypt.GenerateFromPassword(password, bcrypt.DefaultCost)
 	if err != nil {
 		return "", nil, trace.Wrap(err)
 	}
 
-	otp, err := hotp.GenerateHOTP(defaults.HOTPTokenDigits, false)
-	if err != nil {
-		return "", nil, trace.Wrap(err)
-	}
-	hotpQR, err = otp.QR(user)
-	if err != nil {
-		return "", nil, trace.Wrap(err)
-	}
-	hotpURL = otp.URL(user)
+	err = s.UpsertPasswordHash(user, hash)
 	if err != nil {
 		return "", nil, trace.Wrap(err)
 	}
 
-	err = s.UpsertPasswordHash(user, hash)
-	if err != nil {
-		return "", nil, err
-	}
-	err = s.UpsertHOTP(user, otp)
+	otpURL, otpQRCode, err = s.createAndUpsertTOTP(user)
 	if err != nil {
 		return "", nil, trace.Wrap(err)
 	}
 
-	return hotpURL, hotpQR, nil
-
+	return otpURL, otpQRCode, nil
 }
 
-// CheckPassword is called on web user or tsh user login
-func (s *IdentityService) CheckPassword(user string, password []byte, hotpToken string) error {
+// CheckPassword is called on web user or tsh user login using the password and otp token
+func (s *IdentityService) CheckPassword(user string, password []byte, otpToken string) error {
+	var err error
+
 	hash, err := s.GetPasswordHash(user)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+
 	if err := bcrypt.CompareHashAndPassword(hash, password); err != nil {
 		return trace.AccessDenied("passwords do not match")
 	}
-	otp, err := s.GetHOTP(user)
+
+	return s.checkOTP(user, otpToken)
+}
+
+// checkOTP determines the type of OTP token used (for legacy HOTP support), fetches the
+// appropriate type from the backend, and checks the token.
+func (s *IdentityService) checkOTP(user string, otpToken string) error {
```

Comment:
can you please move this logic to `auth` package? Our services are really CRUD applications with no logic. Otherwise I would have to re-implement the same logic when rolling out alternative service.